### ISSUE TITLE
SDK for Route 53 unable to set multiple MX records

### ISSUE THREAD

**Issue by garretwilson** [aws/aws-sdk-java-v2#1484]
I'm trying to create general interface for working with a DNS, and accessing Route 53 through the SDK, I try to perform these common use cases.

* Set multiple `MX` records with different priorities.
* Set multiple `TXT` records (e.g. one for email DKIM, another for validating domain ownership with Atlassian).

Completely unexpectedly, Route 53 causes the last `MX` and last `TXT` record to overwrite all the others.

The description of your [support for `TXT` records](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/ResourceRecordTypes.html#TXTFormat) in the documentation indicates I must put the `TXT` values on separate lines _in a single record_, which is not in compliance with [RFC 1053](https://tools.ietf.org/html/rfc1035). Worse, the documentation for your [`MX` record support](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/ResourceRecordTypes.html#MXFormat) doesn't even mention anything special for multiple values, but a [forum thread](https://forums.aws.amazon.com/thread.jspa?threadID=88634) mentions I must put the `MX` values on separate lines within _the same resource record_. As we [discussed on Server Fault](https://serverfault.com/a/989225/452022), this seems to be not only unexpected but a blatant violation of [RFC 1035](https://tools.ietf.org/html/rfc1035).

Why does Route 53 treat resource records in such a confusing, seemingly non-standards-compliant way? And if you're violating the standards, shouldn't this at least be better documented? Or am I looking in the wrong place?

**Comment by garretwilson** [comment]
Actually combining the `MX` records don't work either. I tried putting `10 smtp1.example.com.` and `20 smtp2.example.com.` on two separate lines as [this thread](https://forums.aws.amazon.com/thread.jspa?threadID=88634) instructs, but I got:

> software.amazon.awssdk.services.route53.model.InvalidChangeBatchException:
> [Invalid Resource Record: FATAL problem: MXRRDATANotTwoFields (MX record doesn't have 2 fields)

So I'm stuck. How do indicate two `MX` records to Route 53 (with different priorities, which is a common use case) using the Java SDK 2? Plus there is still the open question of why handling of multiple `TXT` record values seems to be in contradiction to the standards.

**Comment by garretwilson** [comment]
This is completely blocking my ability to set `MX` multiple records with multiple priorities using the Java SDK 2. I opened a [Stack Overflow question](https://stackoverflow.com/q/58567458/421049) to try to find the problem. Any help is appreciated.

**Comment by garretwilson** [comment]
Today I tried separating the values with `\n`, with `\r\n`, with a comma, and with a space; to no avail.

I even tried setting separate records using `UPSERT` for the first record and `CREATE` for the other records. This failed (as might be expected) with a message indicating that the record already exists. I understand why; I'm just pointing out that I am trying everything!

I'm nearing the conclusion that the SDK is just completely broken here, unless I'm missing something.

**Comment by garretwilson** [comment]
I've _finally_ figured this out. The key is that in the API each name and type is associated with a _resource record set_, that is, `ResourceRecordSet`, and inside it each `ResourceRecord` is only a single _value_. The "resource record set" is has a name+type that functions as a key for the set. In other words, I have to group all my values for the same name and type into this `ResourceRecordSet`, and set them together.

My mistake earlier is that I wrote code to set a single value in a `ResourceRecordSet`, and then I was trying to call that code multiple times, which results in multiple "sets" for the same name+value key. Once I realized this, I had to refactor the code to group the resource records by name+value and put all the values into a `ResourceRecordSet` and use `UPSERT` to set them all at the same time.

So there doesn't seem to be a bug, nor does there seem to be an RFC 1035 violation of specifying multiple values. The values are still specified individually; it's just that they have be grouped together into the same "resource record set", which is a Route 53 specific model peculiarity which confused me after I forgot about that. But nothing seems to be wrong, and it's working, so I'll close this ticket.
